torch/_dynamo/optimizations/distributed.py

```diff
@@ -308,15 +306,28 @@ def run_node(self, n: Node) -> Any:
                         log.debug(
                             f"\n---{n.target} graph---\n" + str(curr_submod.graph)
                         )
+
+                        # When calling the compiler on the submod, inputs (new_args) are expected to
+                        # be FakeTensors already since Dynamo would have made them FakeTensors in the
+                        # non-DDP flow.  However, the parameters are _not_ expected to be FakeTensors,
+                        # since this wrapping happens during compilation
                         compiled_submod_real = self.compile_submod(
                             real_mod, new_args, kwargs
                         )
+
+                        # We update the original (outer) graph with a call into the compiled module
+                        # instead of the uncompiled one.
                         self.module.delete_submodule(n.target)
                         n.target = "compiled_" + n.target
                         self.module.add_submodule(n.target, compiled_submod_real)
-                        return curr_submod(*new_args, **kwargs)
-                    # then we execute the modified node using the usual logic
-                    return getattr(self, n.op)(n.target, new_args, kwargs)
+
+                        # Finally, we have to produce inputs for use compiling the next submodule,
+                        # and these need to be FakeTensors, so we execute the module under fake_mode
+                        with fake_mode:
+                            return curr_submod(*new_args, **kwargs)
```

Comment:
This code is executed at runtime, right?

Reply:
not sure what you mean.

Since torch dynamo defers its compilation until the first execution, then in a way yes, this code happens "at runtime".

But this code only happens as a part of the compilation flow, which in a simple (static model) scenario only happens once.  The second time a user calls their compiled ddp model, none of this code should run, since we're not recompiling.

Reply:
maybe you confused it with 'WrapperModule.forwrad' - that's the only piece of code in the whole `ddp_optimizer' file that I'd expect to run repeatedly on every runtime.  (all it does is unwrap the tuple output from the compiled subgraph)

Reply:
yeah, wrappermodule.forward is the place I was thinking of. 

This looks fine to me.

When Ed and I were working on it - it was very confusing which part of this was compile time, and which was runtime.